group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java

```diff
@@ -590,12 +590,18 @@ private ApiMessage messageOrNull(ApiMessageAndVersion apiMessageAndVersion) {
 
     /**
      * Replays the Record to update the hard state of the group coordinator.
-
-     * @param record The record to apply to the state machine.
+     *
+     * @param producerId    The producer id.
+     * @param producerEpoch The producer epoch.
+     * @param record        The record to apply to the state machine.
      * @throws RuntimeException
      */
     @Override
-    public void replay(Record record) throws RuntimeException {
+    public void replay(
+        long producerId,
```

Comment:
Is the idea here that we pass in the id and the epoch to the relevant records when we implement the offset commit code that can handle them? (But that part is to be implemented)

Reply:
That's right. In the next PR (https://github.com/apache/kafka/pull/14845), we rely on this to build materialize the pending transactional offset commits per producer id.